File: swift-sdk/Sources/GistAdsSDK/Services/DisplayAdAPIService.swift
//
//  DisplayAdAPIService.swift
//  GistAdsSDK
//
//  Service for communicating with the Gist Display Ad API.
//
//  ============================================================================
//  BACKEND CONTRACT (discovered 2026-07-08, PAA-5351 Step 0)
//  ============================================================================
//
//  Discovery method: traced adtag.js's own network calls by downloading and
//  reading the bundle from staging/integration/production, then hit the live
//  endpoint directly with a real demo publisher ID (`guest-api`, found in
//  this SDK's existing test fixtures) to confirm behavior.
//
//  ENDPOINT
//  --------
//  `GET /decision` on a *separate host* from the search API (tp-srch-api):
//    - production:  https://disp-api.prorata.ai
//    - staging:     https://disp-api.staging.prorata.ai
//    - integration: https://prtadsdisplayapi-integration.up.railway.app
//        (temporary Railway infra as of this writing -- NOT the
//        disp-api.integration.prorata.ai pattern used by the other two
//        environments. Overridable via GIST_ADS_DISPLAY_INTEGRATION_URL if
//        this changes; see DisplayAPIConstants.)
//
//  AUTH
//  ----
//  None. There is no Publisher-Key header for display ads. The publisher ID
//  is passed as a plain `publisher` query parameter alongside the page
//  `url`. This makes sense once you see how the browser tag calls this
//  endpoint: via a JSONP `<script>` tag, which cannot carry custom headers.
//  Search ads, by contrast, use a secret Publisher-Key because they're
//  gated more tightly (the native SDK also uses a POST + header for search,
//  a separate code path from the browser's JSONP call to the same
//  endpoint).
//
//  REQUEST
//  -------
//  Query parameters, confirmed directly against the live OpenAPI schema
//  (`GET /openapi.json` on the staging/integration hosts, which expose
//  Swagger docs at `/docs`) -- this superseded and corrected an earlier
//  guess based only on reading adtag.js:
//    - `publisher`:      publisher ID/slug
//    - `url`:            the page URL to target contextually
//    - `ad_size`:        JSON-encoded array, e.g. `[[300,250]]`, or
//                        `[[0,0]]` for dynamic. NOTE: this is `ad_size`,
//                        NOT `sizes` -- an earlier version of this file used
//                        `sizes`, which the backend silently ignores
//                        (unrecognized query params are dropped, not
//                        rejected), so size-based ad selection was
//                        completely inert until this was corrected.
//    - `ad_types`:       pipe-joined type aliases (supported by the backend,
//                        but intentionally NOT sent by this SDK for v1 --
//                        the real aliases, e.g. "generative-image-text",
//                        don't map cleanly onto the existing search-oriented
//                        `AdType` enum, and omitting the param simply
//                        requests all eligible types for the given sizes)
//    - `format`:         "json" (the browser tag uses "jsonp"; this SDK
//                        talks to the endpoint directly, so plain JSON works)
//    - `cb`:             cache-buster (current time in ms) -- not part of
//                        the documented schema, but harmless to include
//    - `correlation_id`: UUID for request tracing -- also not part of the
//                        documented schema, but harmless to include
//    - `data`:           JSON-encoded object of publisher-provided key-value
//                        data, per the live OpenAPI schema: "JSON object
//                        with publisher-provided key-value data for LLM
//                        context". This is the SDK's escape hatch for
//                        native apps: unlike a real webpage, a native
//                        screen has no crawlable HTML for the backend to
//                        analyze via `url` alone, so `context` lets the
//                        host app hand over explicit signal (category,
//                        keywords, section, etc.) instead. The POST
//                        `/decision` endpoint models this as two separate
//                        fields (`context` and `data`), but the GET
//                        endpoint this SDK uses only exposes `data` -- so
//                        this SDK's `context` param is encoded into that
//                        single `data` query parameter.
//    - `geo`:            NOTE: confirmed in the POST `/decision` schema
//                        (`DecisionRequest.geo`, "Optional geographic
//                        location override (country code)") but NOT
//                        present in the GET endpoint's parameter list --
//                        the GET endpoint (used here) has no equivalent
//                        override today, so this SDK does not send `geo`.
//                        If/when the GET endpoint adds it, wire it through
//                        the same way as `data` below.
//
//  RESPONSE
//  --------
//  Raw ad fields, NOT an iframeUrl. Confirmed by reading adtag.js's own
//  `parseResponse` / `convertAdResponseJson`, which normalize and feed
//  fields like `adId`, `adUrl`, `adHeadline`, `adText`, `adCTA`, `adImage`,
//  `adName`, `templateId` into client-side templates -- `iframeUrl` never
//  appears anywhere in the tag's code. The response may be wrapped in
//  `{"selection": [...]}` (matching the search API's convention) or be a
//  flat single-ad object at the root; `DisplayAdResponse` handles both (see
//  that file).
//
//  NOTE: the live schema (`/openapi.json`) uses `adCTA` (all-caps CTA), not
//  `adCta` -- see the `CodingKeys` override in `DisplayAdItem`. A live
//  response also includes several fields not modeled here (`adRelevance`,
//  `adSource`, `flightId`, `brandColors`, `viewUrl`, `adSize`/`width`/
//  `height` echoing back what was actually served, etc.); `Codable` ignores
//  unrecognized keys by default, so decoding is unaffected -- only the v1
//  rendering-relevant fields are modeled.
//
//  NO-FILL
//  -------
//  Confirmed LIVE against staging, integration, and production using the
//  `guest-api` demo publisher ID: every environment returned `HTTP 200`
//  with body `{}` (empty JSON object, no `adId`). No special error status
//  or error envelope -- absence of `adId` is the no-fill signal.
//
//  WHY THIS MEANS A CUSTOM RENDERER
//  ---------------------------------
//  Because the response is raw fields rather than an iframeUrl, this SDK
//  cannot reuse `IframeHTMLGenerator` for display ads. See
//  `DisplayAdHTMLGenerator.swift` for the (intentionally minimal, isolated)
//  raw-field renderer used instead.
//  ============================================================================
//

import Foundation

/// API service for fetching display ads from the Gist Display Ad API.
class DisplayAdAPIService {

    private let baseURL: String
    private let publisherID: String
    private let urlSession: URLSession

    init(baseURL: String, publisherID: String, urlSession: URLSession = .shared) {
        self.baseURL = baseURL
        self.publisherID = publisherID
        self.urlSession = urlSession
    }

    /// Fetch a display ad from the Display Ad API.
    /// - Parameters:
    ///   - pageURL: The current page/context URL to target the ad against.
    ///   - sizes: One or more supported ad sizes.
    ///   - theme: Theme preference - "light" or "dark" (resolved from system if "system" was selected).
    ///   - context: Optional publisher-provided key-value data (e.g. category,
    ///     keywords, section) sent as the `data` query param for LLM context.
    ///     Primarily useful for native apps, where `pageURL` alone gives the
    ///     backend nothing crawlable to analyze -- see the contract notes
    ///     above.
    /// - Returns: HTML string containing the rendered ad.
    /// - Throws: `DisplayAdAPIError.noFill` when the server has no ad to serve.
    func fetchAd(pageURL: String, sizes: [AdSize], theme: String, context: [String: Any]? = nil) async throws -> String {
        let request = try buildRequest(pageURL: pageURL, sizes: sizes, context: context)

        let (data, response) = try await urlSession.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw DisplayAdAPIError.invalidResponse
        }

        guard httpResponse.statusCode == 200 else {
            throw DisplayAdAPIError.httpError(statusCode: httpResponse.statusCode, response: data)
        }

        let decoded: DisplayAdResponse
        do {
            decoded = try JSONDecoder().decode(DisplayAdResponse.self, from: data)
        } catch {
            throw DisplayAdAPIError.invalidData(underlying: error)
        }

        guard let ad = decoded.ad else {
            throw DisplayAdAPIError.noFill
        }

        return DisplayAdHTMLGenerator.generate(ad: ad, theme: theme)
    }

    /// Build the GET request for the `/decision` endpoint.
    /// - Parameter context: Optional publisher-provided key-value data, JSON-encoded into the `data` query param.
    func buildRequest(pageURL: String, sizes: [AdSize], context: [String: Any]? = nil) throws -> URLRequest {
        guard !sizes.isEmpty else {
            throw DisplayAdAPIError.invalidSizes
        }

        guard var components = URLComponents(string: baseURL) else {
            throw DisplayAdAPIError.invalidURL
        }

        let sizesJSON = try AdSize.encodeSizesParam(sizes)
        components.path = components.path.replacingOccurrences(of: "/$", with: "", options: .regularExpression)
            + DisplayAPIConstants.decisionEndpoint
        var queryItems = [
            URLQueryItem(name: "publisher", value: publisherID),
            URLQueryItem(name: "url", value: pageURL),
            URLQueryItem(name: "ad_size", value: sizesJSON),
            URLQueryItem(name: "format", value: "json"),
            URLQueryItem(name: "cb", value: String(Int(Date().timeIntervalSince1970 * 1000))),
            URLQueryItem(name: "correlation_id", value: UUID().uuidString)
        ]

        if let context = context, !context.isEmpty {
            queryItems.append(URLQueryItem(name: "data", value: try Self.encodeContextParam(context)))
        }

        components.queryItems = queryItems

        guard let url = components.url else {
            throw DisplayAdAPIError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        return request
    }

    /// Encode a publisher-provided context dictionary into the JSON string
    /// expected by the `data` query param.
    /// - Throws: `DisplayAdAPIError.invalidContext` if the dictionary contains
    ///   values that aren't valid JSON (e.g. non-JSON types).
    static func encodeContextParam(_ context: [String: Any]) throws -> String {
        guard JSONSerialization.isValidJSONObject(context) else {
            throw DisplayAdAPIError.invalidContext
        }
        let data = try JSONSerialization.data(withJSONObject: context)
        return String(data: data, encoding: .utf8) ?? "{}"
    }
}

/// Errors that can occur during Display Ad API operations
enum DisplayAdAPIError: LocalizedError, Equatable {
    case invalidURL
    case invalidSizes
    case invalidContext
    case invalidResponse
    case invalidData(underlying: Error)
    case httpError(statusCode: Int, response: Data?)
    case noFill

    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid Display Ad API URL"
        case .invalidSizes:
            return "At least one AdSize must be provided"
        case .invalidContext:
            return "Context data must be a valid JSON object (String/Number/Bool/Array/Dictionary values only)"
        case .invalidResponse:
            return "Invalid response from server"
        case .invalidData(let underlying):
            return "Unable to parse response data: \(underlying.localizedDescription)"
        case .httpError(let statusCode, _):
            return "HTTP error: \(statusCode)"
        case .noFill:
            return "No display ad available for this request"
        }
    }

    static func == (lhs: DisplayAdAPIError, rhs: DisplayAdAPIError) -> Bool {
        switch (lhs, rhs) {
        case (.invalidURL, .invalidURL),
             (.invalidSizes, .invalidSizes),
             (.invalidContext, .invalidContext),
             (.invalidResponse, .invalidResponse),
             (.noFill, .noFill):
            return true
        case (.invalidData, .invalidData):
            return true
        case (.httpError(let lStatus, _), .httpError(let rStatus, _)):
            return lStatus == rStatus
        default:
            return false
        }
    }
}

File: android-sdk/src/main/java/com/gist/ads/sdk/services/DisplayAdAPIService.kt
package com.gist.ads.sdk.services

//  Service for communicating with the Gist Display Ad API.
//
//  ============================================================================
//  BACKEND CONTRACT
//  ============================================================================
//
//  ENDPOINT
//  --------
//  `GET /decision` on a *separate host* from the search API (tp-srch-api):
//    - production:  https://disp-api.prorata.ai
//    - staging:     https://disp-api.staging.prorata.ai
//    - integration: https://prtadsdisplayapi-integration.up.railway.app
//        (temporary Railway infra as of this writing -- NOT the
//        disp-api.integration.prorata.ai pattern used by the other two
//        environments. Overridable via the gist.ads.display.integration.url
//        system property if this changes; see DisplayAPIConstants.)
//
//  AUTH
//  ----
//  None. There is no Publisher-Key header for display ads. The publisher ID
//  is passed as a plain `publisher` query parameter alongside the page
//  `url`. This makes sense once you see how the browser tag calls this
//  endpoint: via a JSONP `<script>` tag, which cannot carry custom headers.
//  Search ads, by contrast, use a secret Publisher-Key because they're
//  gated more tightly.
//
//  REQUEST
//  -------
//  Query parameters, confirmed directly against the live OpenAPI schema
//  (`GET /openapi.json` on the staging/integration hosts, which expose
//  Swagger docs at `/docs`):
//    - `publisher`:      publisher ID/slug
//    - `url`:            the page URL to target contextually
//    - `ad_size`:        JSON-encoded array, e.g. `[[300,250]]`, or
//                        `[[0,0]]` for dynamic. NOTE: this is `ad_size`,
//                        NOT `sizes` -- an unrecognized query param is
//                        silently dropped by the backend, not rejected, so
//                        getting this name wrong makes size-based ad
//                        selection completely inert.
//    - `ad_types`:       pipe-joined type aliases (supported by the backend,
//                        but intentionally NOT sent by this SDK for v1 --
//                        omitting the param simply requests all eligible
//                        types for the given sizes)
//    - `format`:         "json" (the browser tag uses "jsonp"; this SDK
//                        talks to the endpoint directly, so plain JSON works)
//    - `cb`:             cache-buster (current time in ms) -- not part of
//                        the documented schema, but harmless to include
//    - `correlation_id`: UUID for request tracing -- also not part of the
//                        documented schema, but harmless to include
//    - `data`:           JSON-encoded object of publisher-provided key-value
//                        data, per the live OpenAPI schema: "JSON object
//                        with publisher-provided key-value data for LLM
//                        context". This is the SDK's escape hatch for
//                        native apps: unlike a real webpage, a native
//                        screen has no crawlable HTML for the backend to
//                        analyze via `url` alone, so `context` lets the
//                        host app hand over explicit signal (category,
//                        keywords, section, etc.) instead. The POST
//                        `/decision` endpoint models this as two separate
//                        fields (`context` and `data`), but the GET
//                        endpoint this SDK uses only exposes `data` -- so
//                        this SDK's `context` param is encoded into that
//                        single `data` query parameter.
//    - `geo`:            confirmed in the POST `/decision` schema but NOT
//                        present in the GET endpoint's parameter list --
//                        so this SDK does not send `geo`. If/when the GET
//                        endpoint adds it, wire it through the same way as
//                        `data` above.
//
//  RESPONSE
//  --------
//  Raw ad fields, NOT an iframeUrl -- see DisplayAdResponse.kt. The response
//  may be wrapped in `{"selection": [...]}` (matching the search API's
//  convention) or be a flat single-ad object at the root; `DisplayAdResponse`
//  handles both.
//
//  NOTE: the live schema (`/openapi.json`) uses `adCTA` (all-caps CTA), not
//  `adCta` -- see the `@SerializedName` override in `DisplayAdItem`.
//
//  NO-FILL
//  -------
//  Confirmed LIVE against staging, integration, and production: every
//  environment returned `HTTP 200` with body `{}` (empty JSON object, no
//  `adId`). No special error status or error envelope -- absence of `adId`
//  is the no-fill signal.
//
//  WHY THIS MEANS A CUSTOM RENDERER
//  ---------------------------------
//  Because the response is raw fields rather than an iframeUrl, this SDK
//  cannot reuse `IframeHTMLGenerator` for display ads. See
//  `DisplayAdHTMLGenerator.kt` for the (intentionally minimal, isolated)
//  raw-field renderer used instead.
//  ============================================================================

import com.gist.ads.sdk.DisplayAPIConstants
import com.gist.ads.sdk.models.AdSize
import com.gist.ads.sdk.models.DisplayAdItem
import com.gist.ads.sdk.models.DisplayAdResponse
import com.gist.ads.sdk.utils.DisplayAdHTMLGenerator
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.IOException
import java.util.UUID
import java.util.concurrent.TimeUnit

/**
 * API service for fetching display ads from the Gist Display Ad API.
 */
class DisplayAdAPIService(
    private val baseUrl: String,
    private val publisherId: String
) {
    private val client: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()

    private val gson: Gson = GsonBuilder()
        .registerTypeAdapter(DisplayAdResponse::class.java, DisplayAdResponse.Deserializer)
        .create()

    /**
     * Fetch a display ad from the Display Ad API.
     * @param pageUrl The current page/context URL to target the ad against.
     * @param sizes One or more supported ad sizes.
     * @param theme Theme preference - "light" or "dark" (resolved from system if "system" was selected).
     * @param context Optional publisher-provided key-value data (e.g. category, keywords, section)
     *   sent as the `data` query param for LLM context. Primarily useful for native apps, where
     *   `pageUrl` alone gives the backend nothing crawlable to analyze -- see the contract notes above.
     * @return HTML string containing the rendered ad.
     * @throws DisplayAdAPIException.NoFill when the server has no ad to serve.
     */
    suspend fun fetchAd(
        pageUrl: String,
        sizes: List<AdSize>,
        theme: String,
        context: Map<String, Any>? = null
    ): String = withContext(Dispatchers.IO) {
        val request = buildRequest(pageUrl, sizes, context)

        val response = try {
            client.newCall(request).execute()
        } catch (e: IOException) {
            throw DisplayAdAPIException.NetworkError(e)
        }

        if (!response.isSuccessful) {
            val body = response.body?.string()
            throw DisplayAdAPIException.HttpError(response.code, body)
        }

        val responseBody = response.body?.string()
            ?: throw DisplayAdAPIException.InvalidResponse

        val decoded: DisplayAdResponse = try {
            gson.fromJson(responseBody, DisplayAdResponse::class.java)
        } catch (e: Exception) {
            throw DisplayAdAPIException.InvalidData(e)
        }

        val ad: DisplayAdItem = decoded.ad ?: throw DisplayAdAPIException.NoFill

        DisplayAdHTMLGenerator.generate(ad, theme)
    }

    /**
     * Build the GET request for the `/decision` endpoint.
     * @param context Optional publisher-provided key-value data, JSON-encoded into the `data` query param.
     */
    internal fun buildRequest(
        pageUrl: String,
        sizes: List<AdSize>,
        context: Map<String, Any>? = null
    ): Request {
        if (sizes.isEmpty()) {
            throw DisplayAdAPIException.InvalidSizes
        }

        val trimmedBase = baseUrl.trimEnd('/')
        val httpUrl = (trimmedBase + DisplayAPIConstants.DECISION_ENDPOINT).toHttpUrlOrNull()
            ?: throw DisplayAdAPIException.InvalidUrl

        val sizesJson = AdSize.encodeSizesParam(sizes)

        val urlBuilder = httpUrl.newBuilder()
            .addQueryParameter("publisher", publisherId)
            .addQueryParameter("url", pageUrl)
            .addQueryParameter("ad_size", sizesJson)
            .addQueryParameter("format", "json")
            .addQueryParameter("cb", System.currentTimeMillis().toString())
            .addQueryParameter("correlation_id", UUID.randomUUID().toString())

        if (!context.isNullOrEmpty()) {
            urlBuilder.addQueryParameter("data", encodeContextParam(context))
        }

        return Request.Builder()
            .url(urlBuilder.build())
            .get()
            .build()
    }

    companion object {
        /**
         * Encode a publisher-provided context map into the JSON string
         * expected by the `data` query param.
         * @throws DisplayAdAPIException.InvalidContext if the map contains
         *   values that aren't valid JSON (e.g. non-JSON types).
         */
        fun encodeContextParam(context: Map<String, Any>): String {
            if (!isValidJsonValue(context)) {
                throw DisplayAdAPIException.InvalidContext
            }
            return Gson().toJson(context)
        }

        private fun isValidJsonValue(value: Any?): Boolean = when (value) {
            null -> true
            is String, is Boolean, is Int, is Long, is Float, is Double, is Short -> true
            is Map<*, *> -> value.keys.all { it is String } && value.values.all { isValidJsonValue(it) }
            is List<*> -> value.all { isValidJsonValue(it) }
            is Array<*> -> value.all { isValidJsonValue(it) }
            else -> false
        }
    }
}

/**
 * Errors that can occur during Display Ad API operations.
 */
sealed class DisplayAdAPIException(message: String, cause: Throwable? = null) : Exception(message, cause) {
    object InvalidUrl : DisplayAdAPIException("Invalid Display Ad API URL")
    object InvalidSizes : DisplayAdAPIException("At least one AdSize must be provided")
    object InvalidContext : DisplayAdAPIException(
        "Context data must be a valid JSON object (String/Number/Boolean/List/Map values only)"
    )
    object InvalidResponse : DisplayAdAPIException("Invalid response from server")

    data class InvalidData(val underlying: Throwable) : DisplayAdAPIException(
        "Unable to parse response data: ${underlying.message}",
        underlying
    )

    data class HttpError(val statusCode: Int, val responseBody: String?) : DisplayAdAPIException(
        "HTTP error: $statusCode"
    )

    object NoFill : DisplayAdAPIException("No display ad available for this request")

    data class NetworkError(val error: Throwable) : DisplayAdAPIException(
        "Network error: ${error.message ?: "Unable to connect to ad server"}",
        error
    )
}
